### PR TITLE
Feature/issues/794 Stats

### PR DESCRIPTION
**Pull request by Mizzick**
#794

### CODE REVIEW

**Comment by Mizzick** [comment]
screenshot 2017-08-14 20 01 42

File: Extension/lib/pages/popup-controller.js [hunk 1]
```diff
@@ -258,14 +265,167 @@ PopupController.prototype = {
...
+                stats.lastMonth.forEach(function (d) {
+                    result.push(d.total);
+                });
+                break;
+            case 'year':
+                stats.lastYear.forEach(function (d) {
+                    result.push(d.total);
+                });
+                break;
+                break;
```
Comment (atropnikov) [inline Extension/lib/pages/popup-controller.js]: unnecessary

File: Extension/lib/utils/page-stats.js [hunk 2]
```diff
@@ -92,10 +92,192 @@ adguard.pageStats = (function (adguard) {
...
+     * Returns statistics data object
+     */
+    var getStatisticsData = function () {
+        var stats = pageStatsHolder.stats;
+        if (!stats) {
+            return null;
+        }
+
+        return {
+            today: stats.data.hours,
```
Comment (atropnikov) [inline Extension/lib/utils/page-stats.js]: `stats.data` may be `undefined` right after install.

File: Extension/lib/utils/page-stats.js [hunk 3]
```diff
@@ -92,10 +92,192 @@ adguard.pageStats = (function (adguard) {
...
+            lastYear: stats.data.months.slice(-12),
+            overall: stats.data.months
+        };
+    };
+
     return {
         resetStats: resetStats,
         updateTotalBlocked: updateTotalBlocked,
-        getTotalBlocked: getTotalBlocked
+        updateStats: updateStats,
```
Comment (atropnikov) [inline Extension/lib/utils/page-stats.js]: Please, add tests.

File: Extension/lib/utils/page-stats.js [hunk 4]
```diff
@@ -92,10 +92,192 @@ adguard.pageStats = (function (adguard) {
...
+
+    var updateStatsData = function (now, type, blocked, current) {
+        var currentDate = new Date(current.updated);
+
+        var result = current;
+
+        if (adguard.utils.dates.isSameHour(now, currentDate) && result.hours.length > 0) {
+            result.hours[result.hours.length - 1] = updateStatsDataItem(type, blocked, result.hours[result.hours.length - 1]);
+        } else {
+            var diffHours = adguard.utils.dates.getDifferenceInHours(now, currentDate);
```
Comment (atropnikov) [inline Extension/lib/utils/page-stats.js]: You can limit `diffHours` with the `MAX_DAYS_HISTORY` value, as well as `diffDays`

**Comment by ameshkov** [comment]
Last day: every 3rd hour
Week: days of week
Last month: every 3rd day
Last year: every month

**Comment by Mizzick** [comment]
screenshot 2017-08-18 13 27 56
screenshot 2017-08-18 13 28 35

Need a bit of styling more please
@kaprielov

File: Extension/lib/utils/page-stats.js (hunk 4 above)
Comment (Mizzick) [inline Extension/lib/utils/page-stats.js]: No, what's the point to do it?

File: Extension/lib/utils/page-stats.js (hunk 2 above)
Comment (Mizzick) [inline Extension/lib/utils/page-stats.js]: yep

File: Extension/lib/utils/page-stats.js (hunk 3 above)
Comment (Mizzick) [inline Extension/lib/utils/page-stats.js]: It's impossible, cause we use localstorage etc

**Comment by atropnikov** [comment]
The extension has been just installed. Let's add more months.
https://www.dropbox.com/s/zywd6qr1lw79kbe/%D0%A1%D0%BA%D1%80%D0%B8%D0%BD%D1%88%D0%BE%D1%82%202017-08-22%2011.00.04.png?dl=0

The tooltip is out of view. Also, it's the strange range of hours. I've created this screenshot at ~ 11 am. I think the range should be [12, 15, 18, 21, 0, 3, 6, 9, 12]
https://www.dropbox.com/s/ag6wfe0csxq1cwl/%D0%A1%D0%BA%D1%80%D0%B8%D0%BD%D1%88%D0%BE%D1%82%202017-08-22%2011.00.56.png?dl=0

The tooltip is out of view again. Maximum on the chart should be at 22nd day.
https://www.dropbox.com/s/088no8ud1ovwloo/%D0%A1%D0%BA%D1%80%D0%B8%D0%BD%D1%88%D0%BE%D1%82%202017-08-22%2011.02.21.png?dl=0

**Pull request closed without merging**